### PR TITLE
HPCC-20149 ESP support for Dynamic ESDL Proxy mode

### PR DESCRIPTION
**Pull request by afishbeck**
Dynamic ESDL should support a mixture of locally configured service
methods and methods that are proxied off to other ESP instances.

This will make the migration of existing environments much easier
to manage.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by afishbeck** [comment]
@wangkx @mayx please review.

File: esp/bindings/http/client/httpclient.cpp [hunk 1]
```diff
@@ -405,13 +405,134 @@ int CHttpClient::sendRequest(const char* method, const char* contenttype, String
...
+void copyHeaders(CHttpMessage &copyTo, CHttpMessage &copyFrom)
+{
+    if (copyFrom.queryHeaders().ordinality())
+    {
+        ForEachItemIn(i, copyFrom.queryHeaders())
+        {
+            StringArray pair;
+            pair.appendList(copyFrom.queryHeaders().item(i), ":", true);
+            const char *name = pair.item(0);
+            const char *value = pair.item(1);
```
Comment (wangkx) [inline esp/bindings/http/client/httpclient.cpp]: Do you need the 'value' here?

File: esp/bindings/http/client/httpclient.cpp [hunk 2]
```diff
@@ -405,13 +405,134 @@ int CHttpClient::sendRequest(const char* method, const char* contenttype, String
...
+
+    Owned<IMultiException> me = MakeMultiException();
+
+    copyCookies(*httprequest, *forwardRequest, m_host);
+
+    httprequest->send();
+
+    if (m_readTimeoutSecs)
+        httpresponse->setTimeOut(m_readTimeoutSecs);
+    httpresponse->receive(false, me);  // MORE - pass in IMultiException if we want to see exceptions (which are not fatal)
```
Comment (wangkx) [inline esp/bindings/http/client/httpclient.cpp]: Do we need to 'forward' the me into forwardResponse?

File: esp/services/esdl_svc_engine/esdl_binding.cpp [hunk 3]
```diff
@@ -1812,7 +1932,31 @@ int EsdlBindingImpl::onGet(CHttpRequest* request, CHttpResponse* response)
...
+                //only proxy actual service method calls, not meta service calls
+                case sub_serv_method:
+                case sub_serv_query:
+                case sub_serv_instant_query:
+                {
+                    StringBuffer method;
+                    nextPathNode(thepath, method);
+                    StringBuffer proxyAddress;
+                    if (checkForMethodProxy(root, method, proxyAddress))
+                        return forwardProxyMessage(proxyAddress, request, response);
```
Comment (wangkx) [inline esp/services/esdl_svc_engine/esdl_binding.cpp]: 'break;'?

**Review by wangkx — COMMENTED** [review]
@afishbeck some minor comments/questions.

**Review by mayx — COMMENTED** [review]
Don't see any issue with the coding. A couple of general comments:
1) In the Proxy section, maybe the @address can be renamed to "@forwardTo" or something like that? It can be confusing as to whether the address is being used a proxy, vs actually the current esp is acting as a proxy to forward to the @address specified. 
2) How can we prevent someone misconfiguring the @address to point back to itself, thus creating a dire situation of endless loop? Maybe can utilize x-forwarded-for or some kind of custom header?

**Comment by mayx** [comment]
Maybe can utilize "x-forward-for" or a custom header?

File: esp/bindings/http/client/httpclient.cpp (hunk 1 above)
Comment (afishbeck) [reply esp/bindings/http/client/httpclient.cpp]: Good catch I use pair.item(1) below so can remove it.

File: esp/bindings/http/client/httpclient.cpp (hunk 2 above)
Comment (afishbeck) [reply esp/bindings/http/client/httpclient.cpp]: Good point. Hmmm I'll have to look into what these represent.  At least make sure they are logged if not forwarded.

File: esp/services/esdl_svc_engine/esdl_binding.cpp (hunk 3 above)
Comment (afishbeck) [reply esp/services/esdl_svc_engine/esdl_binding.cpp]: yes, should add it.

**Comment by afishbeck** [comment]
@mayx agreed will change @ address.  And I should look into adding x-forwarded-for in any case.  Including perhaps to prevent endless loop.

**Comment by afishbeck** [comment]
@afishbeck keep ownership

**Comment by afishbeck** [comment]
@wangkx @mayx updated based on your past comments.  Please review.

**Comment by afishbeck** [comment]
@wangkx @mayx please do final review.

**Comment by wangkx** [comment]
Looks fine for me.

**Review by mayx — APPROVED** [review]
Looks good.

**Pull request merged**